### PR TITLE
Move vector multiply high/low and even/odd to decodetree, implement vmul[eo][su]d

### PR DESCRIPTION
**Pull request by alqotel**
Moved vector multiply high and low instructions and implemented vector multiply high with gvec.
Moved vector multiply even and odd to decodetree and implemented new instruction with doubleword.

### CODE REVIEW

File: target/ppc/translate/vector-impl.c.inc [hunk 1]
```diff
@@ -81,8 +108,116 @@ TRANS_FLAGS_E(ALTIVEC, ISA310, VMULOSD, do_vx_helper, gen_helper_vmulosd)
...
+    REQUIRE_INSNS_FLAGS2(ctx, ISA310);
+    REQUIRE_VECTOR(ctx);
+
+    static const TCGOpcode vecop_list[] = {
+        INDEX_op_mul_vec, INDEX_op_add_vec, INDEX_op_shri_vec, 0
+    };
+
+    static const GVecGen3i op[4] = {
+        {},
+        {},
```
[the remark this answers is not in the document]
Comment (alqotel) [reply target/ppc/translate/vector-impl.c.inc:165]: I was divided between using `&op[vece - 2]`, `&op[vece - MO_32]` and creating empty spaces in the array, so I'll go with `&op[vece - 2]` like you said then.

File: target/ppc/translate/vector-impl.c.inc [hunk 2]
```diff
@@ -81,8 +108,116 @@ TRANS_FLAGS_E(ALTIVEC, ISA310, VMULOSD, do_vx_helper, gen_helper_vmulosd)
...
+    unsigned bits;
+    bits = (imm == MO_32) ? 16 : 32;
+
+    a1 = tcg_temp_new_vec_matching(t);
+    b1 = tcg_temp_new_vec_matching(t);
+    w  = tcg_temp_new_vec_matching(t);
+    k  = tcg_temp_new_vec_matching(t);
+    mask = tcg_temp_new_vec_matching(t);
+
+    tcg_gen_dupi_vec(vece, mask, (imm == MO_32) ? 0xFFFF : 0xFFFFFFFF);
```
[the remark this answers is not in the document]
Comment (lffpires) [reply target/ppc/translate/vector-impl.c.inc]: +1 on the need for commenting this. Is this doing something similar to the way mul64 in host-utils is implementad?
And is this correct for the signed versions of the instructions? (see the implementation of muls64 for the sign adjustments)

More importantly, is this faster than the non-vectorized versions, especially for the 64-bit ones?

Not sure if you tried/compared, but another option would be to vectorize just the 32-bit insns, leaving the 64-bit insns with the non-vectorized implementations, which perform just 2 multiplications. Then, the 32-bit vectorized implementation could leverage tcg_gen_gvec_mul[s]64, by doing:

1. Sign/zero-extend words 1 and 3 of a and b to fully occupy words 0+1 and 2+3 (not sure there are vectorized extension primitives, so you may need to shli + sari)
2. Use the appropriate gvec_mul[s]64 to have the high 32 bits of a.word[1] * b.word[1] in result.word[0] and the high 32 bits of a.word[3] * b.word[3] in result.word[2]
3. Right shift (either with gvec_shri or gvec_sari, depending on signedness) the original a and b in preparation for multiplying the remaining words (which now also fully occupy words 0+1 and 2+3)
4. Use the appropriate gvec_mul[s]64 to have the high 32 bits of a.word[0] * b.word[0] in result.word[0] and the high 32 bits of a.word[2] * b.word[2] in result.word[2]
5. Compose the final result from the values obtained in steps 2 and 4.

Please don't take the above suggestion as a good/valid solution blindly. It's just a thought, and I haven't tested it. :)

[the remark this answers is not in the document]
Comment (lffpires) [reply target/ppc/translate/vector-impl.c.inc]: Or is this too much of a hack? :/

[the remark this answers is not in the document]
Comment (mferst) [reply target/ppc/translate/vector-impl.c.inc]: I guess it's worth a shot. If the performance is better, it shouldn't be too complicated to explain this rational with comments. IIUC, `GVecGen3i` in `do_vx_mulhs` would have a trivial `.fni8` for `MO_64` and a `.fniv` and `.fno` with this logic for `MO_32`.

**Review by mferst — APPROVED** [review]
The code is looking good.
nit: The first commit subject only mention the moved instruction, I'd be nice to mention the implemented ones too, e.g.: "move vmul[eo][su][bhw] to decodetree, implement vmul[eo][su]d".

**Comment by mferst** [comment]
Merged in upstream: https://lists.gnu.org/archive/html/qemu-ppc/2022-03/msg00002.html

**Pull request closed without merging**